csrc/python_frontend/translation.cpp

```diff
@@ -0,0 +1,555 @@
+    return std::all_of(
+        logical_domain_extents.begin(),
+        logical_domain_extents.end(),
+        [&](Val* v) {
+          return v->definition() == nullptr ||
+              map_val_to_fd_index_.count(v) > 0;
+        });
+  }
+
+  // Gather the expressions necessary to create a scalar value.
+  std::vector<Expr*> gatherScalarExpressions(Val* v) {
+    NVF_ERROR(v != nullptr);
+    NVF_ERROR(v->isScalar());
+
+    // short-circuit: v does not have a definition.
+    if (v->definition() == nullptr) {
+      return {};
+    }
+
+    std::vector<Expr*> expression_chain;
+    std::unordered_set<Expr*> visited;
+    std::vector<Expr*> to_visit = {v->definition()};
+    while (!to_visit.empty()) {
+      Expr* e = to_visit.back();
+      to_visit.pop_back();
+
+      expression_chain.push_back(e);
+      visited.insert(e);
+
+      for (Val* input : e->inputs()) {
+        // short-circuit: input does not have a definition.
+        if (input->definition() == nullptr) {
+          continue;
+        }
+
+        // short-circuit: input definition is already visited.
+        if (visited.count(input->definition()) > 0) {
+          continue;
+        }
+
+        to_visit.push_back(input->definition());
+      }
+    }
+    return expression_chain;
+  }
+
+  // Gather the scalar expressions necessary to create the logical domain for a
+  // TensorView.
+  std::vector<Expr*> gatherScalarExpressions(TensorView* tv) {
+    NVF_ERROR(tv != nullptr);
+    std::vector<Expr*> logical_domain_expressions;
+    const std::vector<IterDomain*>& logical_out_domain =
+        tv->domain()->logical();
+    for (IterDomain* id : logical_out_domain) {
+      std::vector<Expr*> extent_definitions =
+          gatherScalarExpressions(id->getMaybeExpandedExtent());
+      logical_domain_expressions.insert(
+          logical_domain_expressions.end(),
+          extent_definitions.begin(),
+          extent_definitions.end());
+    }
+    return logical_domain_expressions;
+  }
+
+  // Check that all of the expression's inputs are defined in FusionDefinition.
+  bool checkExpressionDependencies(Expr* e) {
+    bool check_view_dependency =
+        !e->isA<ViewOp>() || checkViewShapeDependency(e->as<ViewOp>());
+    return check_view_dependency &&
+        std::all_of(e->inputs().begin(), e->inputs().end(), [&](const Val* v) {
+             return map_val_to_fd_index_.count(v) > 0;
+           });
+  }
+
+  void translate() {
+    fd_->setupDefinition();
+
+    // Add Fusion inputs to FusionDefinition
+    for (nvfuser::Val* v : fusion_->inputs()) {
+      OptOutConstDispatch::dispatch(v);
+    }
+
+    // Gather all expressions in CPP Fusion.
+    const std::vector<nvfuser::Expr*> fusion_exprs = fusion_->exprs();
+    std::deque<nvfuser::Expr*> to_visit(
+        fusion_exprs.begin(), fusion_exprs.end());
+
+    // Scalar expressions are not handled by usedMathVals, so gather them
+    // manually.
```

Comment:
Yeah it seems like `fusion_->exprs()` should already have all those scalar expressions in topological order to me since it calls `StmtSort::getExprs()`.

Reply:
cc'ing @rdspring1 in case you missed this one.

Reply:
The scalar expressions are defined for the extents of the IterDomains for a `reshape` TensorView's logical domain.
`StmtSort::getExprs()` only traverses a value's inputs. 

`TV1 [id0, id1, (id2 * id3)] = view(TV0[id0, id1, id2, id3])` --- We don't store the scalar expressions in the `ViewOp` itself, so there is nothing to gather with `StmtSort::getExprs`